src/tike/ptycho/probe.py

```diff
@@ -105,35 +109,111 @@ def update_eigen_probe(R, eigen_probe, weights, β=0.1):
+        weights = weights[..., None, None, None, None]
+        norm_weights = np.linalg.norm(weights[0], axis=-5, keepdims=True)**2
+
+        if np.all(norm_weights[0] == 0):
+            raise ValueError('eigen_probe weights cannot all be zero?')
+
+        # FIXME: What happens when weights is zero!?
+        proj = (np.real(R.conj() * eigen_probe) + weights) / norm_weights
+        return weights, np.mean(
+            R * np.mean(proj, axis=(-2, -1), keepdims=True),
+            axis=-5,
+            keepdims=True,
+        )
+
+    weights, update = (list(a) for a in zip(*comm.pool.map(
+        _get_update,
+        R,
+        eigen_probe,
+        weights,
+    )))
+    update = comm.pool.bcast(comm.pool.reduce_mean(
+        update,
         axis=-5,
-        keepdims=True,
-    )
-    eigen_probe += β * update / np.linalg.norm(
+    ))
+
+    def _get_d(patches, diff, eigen_probe, update, β):
+        eigen_probe += β * update / np.linalg.norm(
+            update,
+            axis=(-2, -1),
+            keepdims=True,
+        )
+        assert np.all(np.isfinite(eigen_probe))
+
+        eigen_probe /= np.linalg.norm(eigen_probe, axis=(-2, -1), keepdims=True)
+
+        # Determine new eigen_weights for the updated eigen probe
+        phi = patches * eigen_probe
+        n = np.mean(
+            np.real(diff * phi.conj()),
+            axis=(-1, -2),
+            keepdims=True,
+        )
+        norm_phi = np.square(np.abs(phi))
+        d = np.mean(norm_phi, axis=(-1, -2), keepdims=True)
+        d_mean = np.mean(d, axis=-5, keepdims=True)
+        return n, d, d_mean
+
+    (n, d, d_mean) = (list(a) for a in zip(*comm.pool.map(
+        _get_d,
+        patches,
+        diff,
+        eigen_probe,
         update,
-        axis=(-2, -1),
-        keepdims=True,
-    )
-    assert np.all(np.isfinite(eigen_probe))
+        β=β,
+    )))
+    d_mean = comm.pool.bcast(comm.pool.reduce_mean(
+        d_mean,
+        axis=-5,
+    ))
 
-    eigen_probe /= np.linalg.norm(eigen_probe, axis=(-2, -1), keepdims=True)
+    def _get_weights_mean(n, d, d_mean, weights):
+        d += 0.1 * d_mean
+
+        weight_update = (n / d).reshape(*weights.shape)
+        assert np.all(np.isfinite(weight_update))
+
+        # (33) The sum of all previous steps constrained to zero-mean
+        weights += weight_update
+        return np.mean(
+            weights,
+            axis=-5,
+            keepdims=True,
+        )
 
-    return eigen_probe
+    weights_mean = list(comm.pool.map(
+        _get_weights_mean,
+        n,
+        d,
+        d_mean,
+        weights,
+    ))
+    weights_mean = comm.pool.bcast(comm.pool.reduce_mean(
+        weights_mean,
+        axis=-5,
+    ))
+
+    def _update_weights(weights, weights_mean):
+        weights -= weights_mean
+        return weights[..., 0, 0, 0, 0]
+
+    weights = list(comm.pool.map(
+        _update_weights,
+        weights,
+        weights_mean,
+    ))
```

Comment:
```suggestion
    weights = comm.pool.map(
        _update_weights,
        weights,
        weights_mean,
    )
```

ThreadPool.map() [returns a list already](https://github.com/tomography/tike/blame/2a16ed37e11fa755b7421eb4a071b707e3588388/src/tike/communicators/pool.py#L127) since #118